Cubical/Functions/Embedding.agda

```diff
@@ -176,78 +194,226 @@ module _ {f : A → B} (retf : hasRetract f) where
-Embedding→Subset (_ , f , isPropFiber) x = fiber f x , isPropFiber x
+Embedding→Subset (_ , f , isEmbeddingF) x = fiber f x , isEmbedding→hasPropFibers isEmbeddingF x
 
 Subset→Embedding : {X : Type ℓ} → ℙ X → Σ[ A ∈ Type ℓ ] (A ↪ X)
-Subset→Embedding {X = X} A = D , fst , ψ
+Subset→Embedding {X = X} A = D , fst , Ψ
  where
   D = Σ[ x ∈ X ] x ∈ A
 
-  ψ : hasPropFibers fst
-  ψ x ((y , y∈A) , y≡x) ((z , z∈A) , z≡x) = ΣPathP (r , q)
-   where
-    p : y ≡ z
-    p = y≡x ∙ sym z≡x
-
-    r : (y , y∈A) ≡ (z , z∈A)
-    r = Σ≡Prop (∈-isProp A) p
-
-    q : PathP (λ i → p i ≡ x) y≡x z≡x
-    q i j = hcomp (λ k → λ { (j = i1) → x
-                           ; (i = i0) → y≡x j
-                           ; (i = i1) → z≡x (~ k ∨ j) })
-                  (y≡x (i ∨ j))
-
+  Ψ : isEmbedding fst
+  Ψ w x = isEmbeddingFstΣProp (∈-isProp A)
 
 Subset→Embedding→Subset : {X : Type ℓ} → section (Embedding→Subset {ℓ} {X}) (Subset→Embedding {ℓ} {X})
 Subset→Embedding→Subset _ = funExt λ x → Σ≡Prop (λ _ → isPropIsProp) (ua (FiberIso.fiberEquiv _ x))
 
 Embedding→Subset→Embedding : {X : Type ℓ} → retract (Embedding→Subset {ℓ} {X}) (Subset→Embedding {ℓ} {X})
-Embedding→Subset→Embedding {ℓ = ℓ} {X = X} (A , f , ψ) = cong (Σ-assoc-≃ .fst) p
- where
- χ = Subset→Embedding (Embedding→Subset (A , f , ψ)) .snd .snd
-
- p : (((Σ[ x ∈ X ] fiber f x) , fst) , χ) ≡ ((A , f) , ψ)
- p = Σ≡Prop (λ _ → hasPropFibersIsProp) (equivToIso (fibrationEquiv X ℓ) .Iso.leftInv (A , f))
+Embedding→Subset→Embedding {ℓ = ℓ} {X = X} (A , f , ψ) =
+  cong (equivFun Σ-assoc-≃) (Σ≡Prop (λ _ → isEmbeddingIsProp) (secEq (fibrationEquiv X ℓ) (A , f)))
 
 Subset≃Embedding : {X : Type ℓ} → ℙ X ≃ (Σ[ A ∈ Type ℓ ] (A ↪ X))
 Subset≃Embedding = isoToEquiv (iso Subset→Embedding Embedding→Subset
                                     Embedding→Subset→Embedding Subset→Embedding→Subset)
 
 Subset≡Embedding : {X : Type ℓ} → ℙ X ≡ (Σ[ A ∈ Type ℓ ] (A ↪ X))
 Subset≡Embedding = ua Subset≃Embedding
+
+isEmbedding→embedsFibersIntoSingl
+  : isEmbedding f
+  → ∀ z → fiber f z ↪ singl z
+isEmbedding→embedsFibersIntoSingl {f = f} isE z = e , isEmbE where
+  e : fiber f z → singl z
+  e x = f (fst x) , sym (snd x)
+
+  isEmbE : isEmbedding e
+  isEmbE u v = goal where
+    -- "adjust" ΣeqCf by trivial equivalences that hold judgementally, which should save compositions
+    Dom′ : ∀ u v → Type _
+    Dom′ u v = Σ[ p ∈    fst u  ≡    fst v  ] PathP (λ i → f (p i) ≡ z) (snd u) (snd v)
+    Cod′ : ∀ u v → Type _
+    Cod′ u v = Σ[ p ∈ f (fst u) ≡ f (fst v) ] PathP (λ i →    p i  ≡ z) (snd u) (snd v)
+    ΣeqCf : Dom′ u v ≃ Cod′ u v
+    ΣeqCf = Σ-cong-equiv-fst (_ , isE _ _)
+
+    dom→ : u ≡ v → Dom′ u v
+    dom→ p = cong fst p , cong snd p
+    dom← : Dom′ u v → u ≡ v
+    dom← p i = p .fst i , p .snd i
+
+    cod→ : e u ≡ e v → Cod′ u v
+    cod→ p = cong fst p , cong (sym ∘ snd) p
+    cod← : Cod′ u v → e u ≡ e v
+    cod← p i = p .fst i , sym (p .snd i)
+
```

Comment:
As commented before, this looks really long for doing almost nothing except shuffling stuff around. Not sure if there's a better way to do this.